rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -475,6 +476,13 @@ void RecorderImpl::subscribe_topic(const rosbag2_storage::TopicMetadata & topic)
   writer_->create_topic(topic);
 
   Rosbag2QoS subscription_qos{subscription_qos_for_topic(topic.name)};
+
+  // For service event topic, avoid receiving the last response message.
+  // TODO(Barry-Xu-2018): is there a better way ?
+  if (rosbag2_cpp::is_service_event_topic(topic.name, topic.type)) {
+    subscription_qos.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
+  }
```

Comment:
Comments from MichaelOrlov

Could you please clarify what was the intention here changing QoS durability for service events?
In particular the "avoid receiving the last response message." is unclear to me why.

Reply:
@MichaelOrlov   

By default, service event publisher uses RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL.  
After testing, I find The number of recorded events is always odd.   
If user use info command to show recorded info, it always shows the response count is bigger than request count.  
```
Service information: Service: /add_two_ints | Type: example_interfaces/srv/AddTwoInts | Request Count: 18 | Response Count: 19 | Serialization Format: cdr
```

So I change QoS to RMW_QOS_POLICY_DURABILITY_VOLATILE to avoid receiving the last response message.

Reply:
@MichaelOrlov 

Since service event publisher set RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL, while I use introspection demo to test, I found the first received msg is always response message. 
So the "info" command always displays a mismatch between requests and response, with the number of responses being one more than the number of requests. This information is a concern as it may confuse users.
```
Service information: Service: /add_two_ints | Type: example_interfaces/srv/AddTwoInts | Request Count: 18 | Response Count: 19 | Serialization Format: cdr
```

After checking, the QoS is set by demo code  
https://github.com/ros2/demos/blob/36a4c0bd6ae0531ad3096806c41efd7062a1cb31/demo_nodes_cpp/src/services/introspection_service.cpp#L126-L127
It uses RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT. For fastdds, the default is TRANSIENT_LOCAL_DURABILITY_QOS.
https://github.com/eProsima/Fast-DDS/blob/63dde5631545fa40203008d32629dba87b14d823/src/cpp/fastdds/publisher/qos/DataWriterQos.cpp#L30

So this is decided by user. I am not sure whether this is a bug.   
Do you think we should force setting RMW_QOS_POLICY_DURABILITY_VOLATILE ?

Reply:
@Barry-Xu-2018 Meant it looks like a buggy behavior if we have one more response count than the request count with the `RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL` settings.  
Need to figure out why we have one extra response with `RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL`.

Reply:
@MichaelOrlov 

>  I meant it looks like a buggy behavior if we have one more response count than the request count with the RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL settings.

Okay. Get it.   
But note that, even if we use RMW_QOS_POLICY_DURABILITY_VOLATILE, the case (more response count than the request count) cannot be avoided completely. It is possible that the first recorded message is a response message.

> Need to figure out why we have one extra response with RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL.

I will add more description why we force setting it.

Reply:
@MichaelOrlov

After further consideration, if we want to completely avoid the case (more response count than the request count), we have to add check while recording. At recording phase, we parse part (ServiceEventInfo) of the message to remove unmatched response message for each client ID. Of course, this brings workload on deserialization.

Reply:
@Barry-Xu-2018 As regards:
> But note that, even if we use RMW_QOS_POLICY_DURABILITY_VOLATILE, the case (more response count than the request count) cannot be avoided completely. It is possible that the first recorded message is a response message.

It sounds like an incorrectly written test. Could you please point out in what tests or places you observe extra response messages?

As regards: 
> if we want to completely avoid the case (more response count than the request count), we have to add check while recording. At recording phase, we parse part (ServiceEventInfo) of the message to remove unmatched response message for each client ID. Of course, this brings workload on deserialization.

We should avoid doing deserialization during recording.

Reply:
@MichaelOrlov 

> It sounds like an incorrectly written test. Could you please point out in what tests or places you observe extra response messages?

I didn't explain it clearly.  
When a service and several clients start working, and users start recording service events by rosbag, there's a possibility that the messages received by rosbag will be response events. This is because you can't guarantee that rosbag will start before the service and clients, so you can't ensure that the first recorded messages will be request messages.

Reply:
@Barry-Xu-2018 As regards:
> This is because you can't guarantee that rosbag will start before the service and clients

Actually, we should be. Please let me know where in the tests this is needed and I might help with it.

Reply:
I don't like the idea that we are going to unconditionally enforce overriding QoS durability for service events to the RMW_QOS_POLICY_DURABILITY_VOLATILE.  
It might be considered as a bug.
I would prefer to have it as is by default. Users can override QoS settings for any topic if they want via the `--qos-profile-overrides-path` argument in the CLI https://github.com/ros2/rosbag2#overriding-qos-profiles

Reply:
@MichaelOrlov 

> Actually, we should be. Please let me know where in the tests this is needed and I might help with it.

The situation I'm referring to is when users are using rosbag to record a service. It's not related to testing. We can't ensure that users start rosbag before starting the service/client.  

> I don't like the idea that we are going to unconditionally enforce overriding QoS durability for service events to the RMW_QOS_POLICY_DURABILITY_VOLATILE.
It might be considered as a bug.
I would prefer to have it as is by default. Users can override QoS settings for any topic if they want via the
--qos-profile-overrides-path argument in the CLI https://github.com/ros2/rosbag2#overriding-qos-profiles

After re-considering this, I agree with you.
We should not override QoS durability.  
```c++
  void
  configure_introspection(
    Clock::SharedPtr clock, const QoS & qos_service_event_pub,
    rcl_service_introspection_state_t introspection_state)
```
The QoS can be specified by user.  Even if QoS settings (with RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL) can lead to a mismatch between the number of responses and requests, this is a true reflection of the recorded results. The "info" command accurately displays the recorded number of responses and requests. Even if the response and request counts don't match, it doesn't affect the "play" operation. "play" will still correctly send out all recorded requests.
So I want to remove unconditionally enforce overriding QoS durability for QoS settings and keep current behavior (Maybe mismatch between the number of responses and requests).

Reply:
@Barry-Xu-2018 I see that you agreed to remove QoS durability overriding for service events, but I don't see that you actually did it. It seems it was forgotten.